wallet: improve IBD sync time by skipping block scanning prior birth time

furszy opened the pull request:
During initial block download, the node's wallet(s) scans every arriving block looking for data that it owns.
This process can be resource-intensive, as it involves sequentially scanning all transactions within each
arriving block.

To avoid wasting processing power, we can skip blocks that occurred before the wallet's creation time, 
since these blocks are guaranteed not to contain any relevant wallet data.

This has direct implications (an speed improvement) on the underlying blockchain synchronization process
as well. The reason is that the validation interface queue is limited to 10 tasks per time. This means that no
more than 10 blocks can be waiting for the wallet(s) to be processed while we are synchronizing the chain
(activating the best chain to be more precise).
Which can be a bottleneck if blocks arrive and are processed faster from the network than what they are 
processed by the wallet(s).

So, by skipping not relevant blocks in the wallet's IBD scanning process, we will also improve the chain 
synchronization time.

furszy committed:
wallet: skip block scan if block was created before wallet birthday

To avoid wasting processing power, we can skip blocks that occurred
before the wallet's creation time,  since these blocks are guaranteed
not to contain any relevant wallet data.

This has direct implications (an speed improvement) on the underlying
blockchain synchronization process as well.

The reason is that the validation interface queue is limited to
10 tasks per time. This means that no more than 10 blocks can be
waiting for the wallet(s) to be processed while we are synchronizing
the chain (activating the best chain to be more precise).
Which can be a bottleneck if blocks arrive and are processed faster
from the network than what they are  processed by the wallet(s).

pinheadmz commented:
concept ACK, seems like a great idea.

pinheadmz reviewed (commented):
Did some general code review. I see we already have `nCreateTime` for descriptors and part of this PR seems to be just bringing that up to the wallet level, so that's not a redundancy. 

Also I don't know if this is a problematic edge case or it could just be a good way to test but you could create a key and send BTC to it in a block, then a few blocks later import into a new descriptor and rescan. The wallet *should* miss the first transaction.

furszy commented on src/interfaces/chain.h:
```diff
@@ -83,6 +83,7 @@ struct BlockInfo {
     const uint256& hash;
     const uint256* prev_hash = nullptr;
     int height = -1;
+    int64_t time = 0;
```
Mainly to be consistent with the chain interface, wallet and spkm sources. Timestamps are represented as `int64_t` there. E.g. the `FoundBlock` class, `findFirstBlockWithTimeAndHeight` method, the `CWallet::m_best_block_time` member, first key time, descriptor creation time etc.

I think that is clear that block times are always in seconds, so the use of an arithmetic type shouldn't be much of a worry but we could switch all of them to `NodeSeconds` too (probably in another PR that does it all at once?).

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1777,6 +1784,14 @@ bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScri
+void CWallet::FirstKeyTimeChanged(const ScriptPubKeyMan* spkm, int64_t new_birth_time)
+{
+    int64_t birthtime = m_birth_time.load();
+    if (birthtime == BLANK_BIRTH_TIME || new_birth_time < birthtime) {
+        m_birth_time = new_birth_time;
+    }
```
Because the other one is a return statement (the flow breaks there) while this one is a setter.
Plus, when the line is too long, like this one, I think that braces help readability.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1434,6 +1434,13 @@ void CWallet::blockConnected(const interfaces::BlockInfo& block)
+    // No need to scan block if block was created before the
+    // wallet birthday (as adjusted for block time variability)
+    const int64_t birth_time = m_birth_time.load();
+    if (birth_time == BLANK_BIRTH_TIME || block.time < birth_time - TIMESTAMP_WINDOW) return;
```
yes. `BLANK_BIRTH_TIME` is for blank wallets. Blank wallets are wallets with no keys (no spkm exist). So it's ok to skip blocks as the wallet isn't going to find anything on them.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3565,17 +3592,18 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
     for (const auto& spk_man : GetActiveScriptPubKeyMans()) {
         spk_man->NotifyWatchonlyChanged.connect(NotifyWatchonlyChanged);
         spk_man->NotifyCanGetAddressesChanged.connect(NotifyCanGetAddressesChanged);
+        spk_man->NotifyFirstKeyTimeChanged.connect(std::bind(&CWallet::FirstKeyTimeChanged, this, std::placeholders::_1, std::placeholders::_2));
     }
 }
 
 void CWallet::LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc)
 {
     if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
         auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new ExternalSignerScriptPubKeyMan(*this, desc, m_keypool_size));
-        m_spk_managers[id] = std::move(spk_manager);
+        AddScriptPubKeyMan(std::move(spk_manager));
     } else {
         auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new DescriptorScriptPubKeyMan(*this, desc, m_keypool_size));
-        m_spk_managers[id] = std::move(spk_manager);
+        AddScriptPubKeyMan(std::move(spk_manager));
```
The id is the db lookup key for the descriptor, and also a way to detect db corruption.

We could clean it up from here, yeah. And probably, depending on its usage, we should also think about caching it.

furszy replied:
Ended up re-adding the id arg. So we don't calculate the spkm id twice.

furszy committed:
refactor: single method to append new spkm to the wallet

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3536,6 +3555,14 @@ LegacyScriptPubKeyMan* CWallet::GetOrCreateLegacyScriptPubKeyMan()
     return GetLegacyScriptPubKeyMan();
 }
 
+void CWallet::AddScriptPubKeyMan(std::unique_ptr<ScriptPubKeyMan> spkm_man)
+{
+    const auto& spkm = m_spk_managers[spkm_man->GetID()] = std::move(spkm_man);
+
+    // Update birth time if needed
+    FirstKeyTimeChanged(spkm.get(), spkm->GetTimeFirstKey());
+}
+
```
sure, pushed.

pablomartin4btc reviewed (commented):
Concept ACK. Light code review, I'd to spend more time with it, which I'll do soon. Good stuff!

achow101 commented on src/interfaces/chain.h:
```diff
@@ -83,6 +83,7 @@ struct BlockInfo {
     const uint256& hash;
     const uint256* prev_hash = nullptr;
     int height = -1;
+    int64_t time = 0;
```
In 1a9a349a026af2865e5446091f1e072b7a845819 "wallet: skip block scan if block was created before wallet birthday"

`CBlock` already has a `GetBlockTime()`, so I don't think it's necessary to include the time as a separate field here.

However I'm not sure that we should be using the block's timestamp for this time rather than MTP. With the block timestamp, we could ignore a more recent block that has a timestamp older than its parent which we wouldn't ignore. That would be kinda bad.

furszy replied:
> CBlock already has a GetBlockTime(), so I don't think it's necessary to include the time as a separate field here.

Yeah, that works here. The new field was because I initially implemented this on the filters-only sync branch. Where I have no blocks, only headers and filters.

> However I'm not sure that we should be using the block's timestamp for this time rather than MTP. With the block timestamp, we could ignore a more recent block that has a timestamp older than its parent which we wouldn't ignore. That would be kinda bad.

Hmm, wasn't the timestamp window created to extend the grace period enough so we can use the block time directly?.
In other words, MTP moves the block time an hour back, and we have a grace period of two hours. If the two hours period concerns you, to not be forced to calculate the MTP for every block, what about doubling the grace period for IBD scanning?.
Normally, it would just mean checking ~12 more blocks.

achow101 replied:
> Hmm, wasn't the timestamp window created to extend the grace period enough so we can use the block time directly?.

AFAIK it exists for accepting blocks with future timestamps, and we only use it in the wallet because it's convenient.

But my comment isn't really about the accuracy of the timestamp, but rather about the possibility of skipping a block after we have already started scanning. A block's timestamp can be before its parent. The code as written would let us skip scanning a block even if we had scanned its parent. The reason to use MTP is because it can only move forward. The MTP of a block can never be less than the MTP of its parent, it can be equal, but that's fine for us. In `RescanFromTime`, we find blocks using a different metric, `GetBlockTimeMax()`. This is calculated as the maximum of the current block's timestamp and it's parent's timestamp, so this too also can only move forward. If MTP is too slow to calculate, we could use `GetBlockTimeMax()` instead as it has that same property.

furszy replied:
> But my comment isn't really about the accuracy of the timestamp, but rather about the possibility of skipping a block after we have already started scanning. A block's timestamp can be before its parent. The code as written would let us skip scanning a block even if we had scanned its parent. The reason to use MTP is because it can only move forward. The MTP of a block can never be less than the MTP of its parent, it can be equal, but that's fine for us.

Yeah, where I was going is that we really don't care about those possible skipped blocks if the grace period is large enough. We know that the wallet will not find anything there, it's just to ensure that all blocks created 2 hours before the wallet birth time will be scanned.
In other words,  The idea was to extend the window enough to not worry about unordered timestamps. E.g. going to the extreme, even if the grace period is one or two days, is much better than scanning the entire chain as we do now.

But.. point taken. The `GetBlockTimeMax` idea sounds nicer. It doesn't require any extra calculation like MTP, nor grace period extension (which makes the wallet scan extra blocks). Just add another member into the `BlockInfo` struct and done.

achow101 commented on src/wallet/wallet.h:
```diff
@@ -294,6 +294,11 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
+    static const int64_t BLANK_BIRTH_TIME = -1;
+    // First created key time. Used to skip blocks prior to this time.
+    // 'BLANK_BIRTH_TIME' if wallet is blank.
+    std::atomic<int64_t> m_birth_time{BLANK_BIRTH_TIME};
```
In 1a9a349a026af2865e5446091f1e072b7a845819 "wallet: skip block scan if block was created before wallet birthday"

`m_birth_time` could be initialized with `std::numeric_limits<in64_t>::max()` which would remove the need to compare against `BLANK_BIRTH_TIME`.

furszy replied:
pushed thanks

furszy commented:
Updated per feedback, thanks achow.

Changes:
* Replaced `BLANK_BIRTH_TIME` constant by setting the `m_birth_time` to its max numeric value.
* Tackled a block time variability concern by using the maximum time in the chain (a moving forward only timestamp) instead of the raw block time.
* Extended the grace period to be twice as big as the one used in the rescan process. Even when this is not strictly needed, the rationale is to be a bit more conservative during IBD scanning because it's an automatic process with no user interaction (unlike the rescan process that is usually triggered by an user action such an import).

ishaanam reviewed (commented):
Approach ACK, this looks good so far but I haven't reviewed the first commit yet.

furszy commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2732,10 +2734,13 @@ void DescriptorScriptPubKeyMan::UpdateWalletDescriptor(WalletDescriptor& descrip
         throw std::runtime_error(std::string(__func__) + ": " + error);
     }
 
+    bool signal_birth_time = descriptor.creation_time < m_wallet_descriptor.creation_time;
     m_map_pubkeys.clear();
     m_map_script_pub_keys.clear();
     m_max_cached_index = -1;
     m_wallet_descriptor = descriptor;
+
+    if (signal_birth_time) NotifyFirstKeyTimeChanged(this, m_wallet_descriptor.creation_time);
```
To save some extra processing, because they are performed at different levels.
At the spkm level (here), we are at the lowest one. So, before emitting an event that will require to be processed by all the listeners (the wallets and/or other registered objects), we can just check if the updated descriptor timestamp is lower than the existent one.

furszy commented:
Updated per feedback, thanks @ishaanam.

furszy commented on src/wallet/wallet.h:
```diff
@@ -638,6 +642,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     bool ImportPubKeys(const std::vector<CKeyID>& ordered_pubkeys, const std::map<CKeyID, CPubKey>& pubkey_map, const std::map<CKeyID, std::pair<CPubKey, KeyOriginInfo>>& key_origins, const bool add_keypool, const bool internal, const int64_t timestamp) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     bool ImportScriptPubKeys(const std::string& label, const std::set<CScript>& script_pub_keys, const bool have_solving_data, const bool apply_label, const int64_t timestamp) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
+    /** Updates wallet birth time if 'new_birth_time' is lower */
```
pushed

ishaanam commented:
cr-ACK 76396ca376188631ba46bd47b134881efcc6f755

maflcko reviewed (commented):
left some comments/nits/questions. Feel free to ignore.

furszy commented on src/test/util/mining.cpp:
```diff
@@ -82,7 +82,8 @@ std::shared_ptr<CBlock> PrepareBlock(const NodeContext& node, const CScript& coi
     LOCK(cs_main);
-    block->nTime = Assert(node.chainman)->ActiveChain().Tip()->GetMedianTimePast() + 1;
+    const CBlockIndex* tip = Assert(node.chainman)->ActiveChain().Tip();
+    block->nTime = tip->nHeight == 0 ? GetTime<std::chrono::seconds>().count() : tip->GetBlockTime() + 1;
```
> Is there a reason for the first commit?

It is because this is the only place where we set the chain timestamps to be in the past. In the form of
```
block_1_timestamp = genesis + 1
block_2_timestamp = median(genesis, block_1) + 1
block_3_timestamp = median(genesis, block_1, block_2) + 1
```
And in the wallet, we use the clock time to set the descriptors/keys creation time.

So, even when we create a wallet prior mining the chain, the wallet will always have a birth time after it. Which will lead to the wallet skipping those blocks during sync. This happens on the `wallet_balance.cpp` benchmark.

But.. I just went deeper over it and.. surprise surprise, seems that the functions that create blocks in the past: `generatetoaddress` (which is not the RPC command) and `PrepareBlock` are only used in benchmarks: `wallet_balance.cpp` and `block_assemble.cpp`.. so we can actually re-work this stuff differently cleaning some stuff.

The rest of the code uses `BlockAssembler::CreateNewBlock`, which sets the block time to the current, or mocked, clock time. So, no functional, unit or fuzz tests should be affected by this change.

furszy replied:
Ended up pushing a simpler approach. Focused on correcting `wallet_balance.cpp` only.

In the future, we could generate a fake chain instead of creating a "real" one. Similar to how it's done
in the `wallet_create_tx.cpp` benchmark.

furszy replied:
ha, loved it. I was so deep that ended up complicating things for no reason.
Dropped the first commit and squashed the mock time change into the last one.

furszy commented on src/test/fuzz/process_messages.cpp:
```diff
@@ -39,7 +39,9 @@ FUZZ_TARGET_INIT(process_messages, initialize_process_messages)
 
     ConnmanTestMsg& connman = *static_cast<ConnmanTestMsg*>(g_setup->m_node.connman.get());
     TestChainState& chainstate = *static_cast<TestChainState*>(&g_setup->m_node.chainman->ActiveChainstate());
-    SetMockTime(1610000000); // any time to successfully reset ibd
```
not anymore. Re-worked the first commit so it only touches the `wallet_balance.cpp` bench file and nothing else.

maflcko reviewed (approved):
Concept ACK. Haven't reviewed the code.

furszy commented:
> sanity check: because we already skip old blocks during rescan?
>
> https://github.com/bitcoin/bitcoin/pull/26679

Yes. Different entry points.

#26679 addressed the scanning of the existing chain during load time.
And here, the emphasis shifts to the synchronization process when the
wallet is active.
When the node activate the best chain on each processed block and
signal the "block connected" event.

pinheadmz commented:
To review this PR I wrote a simple functional test: https://github.com/pinheadmz/bitcoin/commit/1f0f945aca3a4f5118e49b2d31318ddd4b19d906 which surprised me by failing. I added logging and comments to the test to share it because I don't entirely understand what's happening but it looks to me like, the wallet generates a `combo(...)` descriptor with a timestamp of `1` and therefore the wallet creation time is always `1` and so the new logic doesn't actually skip any blocks. But if that's the case, does that also mean #26679 doesn't work either?

test output:

```
2023-05-25T23:10:29.244000Z TestFramework (INFO): PRNG seed is: 8288012825723617302
2023-05-25T23:10:29.245000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test__5pngcq0
2023-05-25T23:10:29.886000Z TestFramework (INFO): Generate 10 blocks with 10-hour timestamp gaps
2023-05-25T23:10:29.927000Z TestFramework (INFO): Initial blockchain sync the wallet's node

not(500.00000000 == 250)

Why does this descriptor have a timestamp of 1?
{'desc': 'combo(022d715e6f0d7c037dbccd3327f0b3bc3c76258a7e017e1a64902f76406c87dfda)#pesd53nq', 'timestamp': 1, 'active': False}

2023-05-25T23:10:31.083000Z TestFramework (INFO): Stopping nodes
2023-05-25T23:10:31.306000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test__5pngcq0
2023-05-25T23:10:31.306000Z TestFramework (INFO): Tests successful
```

and then in bitcoin `debug.log`, this output from `CWallet::blockConnected()`: `m_birth_time.load(): 1`. Every time we run that function, the log says wallet birthday is `1`

furszy commented:
> To review this PR I wrote a simple functional test: https://github.com/pinheadmz/bitcoin/commit/1f0f945aca3a4f5118e49b2d31318ddd4b19d906 which surprised me by failing. I added logging and comments to the test to share it because I don't entirely understand what's happening but it looks to me like, the wallet generates a combo(...) descriptor with a timestamp of 1 and therefore the wallet creation time is always 1 and so the new logic doesn't actually skip any blocks. But if that's the case, does that also mean https://github.com/bitcoin/bitcoin/pull/26679 doesn't work either?

The issue is on your test setup. The `combo` descriptor comes from the test framework default wallet creation process, not the inner wallet.

The framework, after creating the default wallet, imports one of the hardcoded private keys ([link](https://github.com/bitcoin/bitcoin/blob/25202cace9140870c75cb3a811e10045df88c226/test/functional/test_framework/test_framework.py#L441)) used to receive the coinbases coins ([link](https://github.com/bitcoin/bitcoin/blob/25202cace9140870c75cb3a811e10045df88c226/test/functional/test_framework/test_node.py#L309)).

This means calling the test framework `importprivkey`, which is a proxy function with two possible outcomes:

1) For legacy wallet, call the RPC command `importprivkey`: which doesn't receive a timestamp, so creation time is initialized to the minimum one.
2) For descriptors wallet, call the RPC command `importdescriptors` with [timestamp=0](https://github.com/bitcoin/bitcoin/blob/25202cace9140870c75cb3a811e10045df88c226/test/functional/test_framework/test_node.py#L787). (which then is changed by the node internally to 1, because of the [minimum supported timestamp](https://github.com/bitcoin/bitcoin/blob/25202cace9140870c75cb3a811e10045df88c226/src/wallet/rpc/backup.cpp#L1669))

The simplest solution for your test is to create the wallet by yourself, by calling `createwallet`, instead of using the default one initialized by the test framework.

pinheadmz commented:
@furszy awwwwwesome thank you for catching that. Ok, passing test is now at https://github.com/pinheadmz/bitcoin/commit/f1238700a1f3d3e88b7f20bc51f0088783d53595

pinheadmz reviewed (approved):
ACK 82bb7831fa6052620998c7eef47e48ed594248a8

Reviewed code and tested locally, wrote a test. I am running this branch side-by-side with master on a VPS and will report who wins the race in a day or so ;-)

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 82bb7831fa6052620998c7eef47e48ed594248a8
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmRwxNIACgkQ5+KYS2KJ
yTq5nhAA3dJDzRO4VmDUA8n9YO4RtpcQBHT/t9NOsoyNFszTZOQb47JqCTDSGBeZ
mwgdfTIMWDkIesi5lG81ATdGTvxeJjx/7OjN90x46Lj3bJuyVTJxrnrVtLuv/2tg
G9fnfXO2Ngod/daq9psvNhLfQLhW3CQNtsJ3x4NOJafKLOyxmd+dKxnXzJaKAicp
ZlnIyUtArxLiDFXRmJRpmKAX2TRj49HooiY6gKlyyTtiOitB48AQdsh6hzaQ2O3q
wyYzIwM8jEC3/hIbemPaoLphc6esevgXRCcEo3zv+VvnDdrtNBwXEKFtThkPniEq
5hF2zaVCB6cqlvjcFMnkZOVafxE9GWe7fJR7OPa0W2vS3YaSZD7zdhwTpK2KV0Cj
HG21b2x5VFep6cQDl1Xqlktu21ExAWg2zOwESCnrJ7riRXtb3+Xe/eny90aeYhyX
0+ORbVy8AVyBYC1+nBjqpJA8JAoncm3xgNTpbgvX8XrflSaLcbhod5UAh0R+NE4c
QGvGDa6ITO0+Cu/Yba4TwgOxzBmA/oXSxUecGR5eQi3KuKobVlfdLbKgjt1tfscs
t414zb/cjcgpxVMXrrLZMMttg06I9YvSVUEcPCeU2YgiLgigTqunwsZ7KsOJM5ih
HB4+y7/lIg+7kme6ZheZ221r7nWY6atR5piFe09AqLmKPBEviio=
=VjDW
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1436,6 +1436,12 @@ void CWallet::blockConnected(const interfaces::BlockInfo& block)
+
+    // No need to scan block if it was created before the wallet birthday.
+    // Uses chain max time and twice the grace period to adjust time for block time variability.
+    if (block.chain_time_max < m_birth_time.load() - (TIMESTAMP_WINDOW * 2)) return;
```
It's merely about consistency and readability.

The explicit `load` call tells the reader that this is an atomic variable, which could prevent some issues. E.g.
```c++
if (block.chain_time_max < m_birth_time - (TIMESTAMP_WINDOW * 2)) return; 
else (m_birth_time > something_esle) // do something else..
```
Here there are two atomic reads performed. And the `m_birth_time` could have changed in-between them.

ishaanam commented:
re-ACK 82bb7831fa6052620998c7eef47e48ed594248a8

achow101 commented:
re-ACK 82bb7831fa6052620998c7eef47e48ed594248a8

furszy commented:
Thanks Marco. Fixed in #27786.
Would be nice if the CI alert us about this kind of stuff prior merging the PR.
The base fuzzer corpus should have detected it right?.
